homeassistant/components/neevo/__init__.py

```diff
@@ -0,0 +1,127 @@
+"""The Nee-Vo Tank Monitoring integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from aiohttp.client_exceptions import ClientError
+from pyneevo import NeeVoApiInterface
+from pyneevo.errors import (
+    GenericHTTPError,
+    InvalidCredentialsError,
+    InvalidResponseFormat,
+    PyNeeVoError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_connect, dispatcher_send
+from homeassistant.helpers.entity import DeviceInfo, Entity
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+
+from .const import API_CLIENT, DOMAIN, TANKS
+
+_LOGGER = logging.getLogger(__name__)
+
+PUSH_UPDATE = "neevo.push_update"
+INTERVAL = timedelta(minutes=15)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Nee-Vo component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][API_CLIENT] = {}
+    hass.data[DOMAIN][TANKS] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up Nee-Vo Tank Monitoring from a config entry."""
+
+    email = config_entry.data[CONF_EMAIL]
+    password = config_entry.data[CONF_PASSWORD]
+
+    try:
+        api = await NeeVoApiInterface.login(email, password=password)
+    except InvalidCredentialsError:
+        _LOGGER.error("Invalid credentials provided")
+        return False
+    except PyNeeVoError as err:
+        _LOGGER.error("Config entry failed: %s", err)
+        raise ConfigEntryNotReady from err
+
+    try:
+        tanks = await api.get_tanks_info()
+    except (ClientError, GenericHTTPError, InvalidResponseFormat) as err:
+        raise ConfigEntryNotReady from err
+
+    hass.data[DOMAIN][API_CLIENT][config_entry.entry_id] = api
+    hass.data[DOMAIN][TANKS][config_entry.entry_id] = tanks
+
+    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
+
+    def update_published():
+        """Handle a push update."""
+        dispatcher_send(hass, PUSH_UPDATE)
+
+    for _tank in tanks.values():
+        _tank.set_update_callback(update_published)
+
+    async def fetch_update(now):
+        """Fetch the latest changes from the API."""
+        await api.refresh_tanks()
+
+    config_entry.async_on_unload(
+        async_track_time_interval(hass, fetch_update, INTERVAL + timedelta(minutes=1))
```

Comment:
I think I forgot that in there from an integration I was using as a reference. That integrations API updated on the hour, so if you didn't add a bit onto it, there was the possibility it wouldn't update. Yeah, I should probably have it on the hour. In my case, our generator is running off our tank so in an emergency/when we are running on the generator, it would be nice to have it as live as can be. Maybe I could add a switch to increase the polling rate. Something you could do either manually or automatically based on if the generator is running or not.

Reply:
IDK if you would have any idea of the best way to implement that switch. If you do would you be able to point me in the right direction?

Reply:
I'd probably leave that up to a user authored automation to call the `homeassistant.update_entity` service on a timer when in generator power mode to force updates to happen more often. I've never used that service and I don't know how it would interact with what you have now or with a DataUpdateCoordinator.